includes/classes/Product.php

```diff
@@ -0,0 +1,306 @@
+        $sql = "SELECT products_id FROM " . TABLE_SPECIALS . " WHERE products_id=" . (int)$this->product_id;
+        $results = $db->Execute($sql, 1);
+        return !$results->EOF;
+    }
+
+    public function priceIsByAttribute(): bool
+    {
+        return ($this->data['products_priced_by_attribute'] ?? '0') === '1';
+    }
+
+    public function priceIsFree(): bool
+    {
+        return ($this->data['product_is_free'] ?? '0') === '1';
+    }
+
+    public function priceIsCall(): bool
+    {
+        return ($this->data['product_is_call'] ?? '0') === '1';
+    }
+
+    public function __get(string $name)
+    {
+        return $this->get($name);
+    }
+
+    protected function loadProductDetails(int $product_id, ?int $language_id = null): array
+    {
+        global $db;
+
+        $sql = "SELECT p.*, pt.allow_add_to_cart, pt.type_handler, m.manufacturers_name, m.manufacturers_image
+                FROM " . TABLE_PRODUCTS . " p
+                LEFT JOIN " . TABLE_PRODUCT_TYPES . " pt ON (p.products_type = pt.type_id)
+                LEFT JOIN " . TABLE_MANUFACTURERS . " m USING (manufacturers_id)
+                WHERE p.products_id = " . (int)$product_id;
+        $product = $db->Execute($sql, 1, true, 900);
+
+        if ($product->EOF) {
+            return [];
+        }
+
+        $data = $product->fields;
+        $data['id'] = $data['products_id'];
+        $data['product_id'] = $data['products_id'];
+        $data['info_page'] = $data['type_handler'] . '_info';
+        //$data['parent_category_id'] = $data['master_categories_id'];
+
+        /**
+         * Add $data['lang'][code] = [products_name, products_description, etc] for each language
+         */
+        $sql = "SELECT pd.*
+                FROM " . TABLE_PRODUCTS_DESCRIPTION . " pd
+                WHERE pd.products_id = " . (int)$product_id . "
+                ORDER BY language_id";
+        $pd = $db->Execute($sql, null, true, 900);
+        foreach ($pd as $result) {
+            unset($result['products_id']);
+            $data['lang'][$this->languages[$result['language_id']]] = $result;
+        }
+
+        // count linked categories
+        $sql = "SELECT categories_id FROM " . TABLE_PRODUCTS_TO_CATEGORIES . " ptc WHERE products_id=" . (int)$product_id;
+        $results = $db->Execute($sql, null, true, 900);
+        $data['linked_categories_count'] = $results->RecordCount();
+        $data['linked_categories'] = [];
+        foreach($results as $result) {
+            $data['linked_categories'][] = $result['categories_id'];
+        }
+
+        // get cPath
+        $categories = [];
+        zen_get_parent_categories($categories, $data['master_categories_id']);
+        $categories = array_reverse($categories);
+        $categories[] = $data['master_categories_id'];
+        $data['cPath'] = implode('_', $categories);
+
+
+        //Allow an observer to modify details
+        $this->notify('NOTIFY_GET_PRODUCT_OBJECT_DETAILS', $product_id, $data);
```

Comment:
Well, the data being passed is in a very different format.
The pass-by-reference `$data` here is an array of actual final data ... which IMO is the most intuitive, and flexible (as opposed to the following).
The pass-by-reference `$product` in `NOTIFY_GET_PRODUCT_DETAILS` is a `queryFactoryResult` object, via which you must access the `->fields` sub-array to work with the passed data.

I had chosen a new notifier hook name to avoid the clash of unexpected data formats. But if you're okay with the incoming data changing (which will break all un-updated observers formerly watching for `NOTIFY_GET_PRODUCT_DETAILS`, then I agree we can re-use the old notifier name, with the new $data passed to it.

Reply:
Crud.  While I'm totally in support of this class' creation, this is going to be a goat-rodeo.

Reply:
I guess if the old one disappears, then only the new one would fire ... so that's one way to avoid a clash.

Or, we could reuse the old name and just instruct everyone to update their use of the old notifier name to check for whether the `[fields]` sub-array exists, and take action accordingly.

Reply:
Either way, it's a backward-compatibility cluster.

Reply:
Could we make it backwards compatible by just doing the old notifier above this with old parameters? 

```
// Backwards compatibility; new plugins should use NOTIFY_GET_PRODUCT_OBJECT_DETAILS instead
$old_product = $product;
$old_product->fields = array_merge($old_product->fields, $pd->fields);
 $zco_notifier->notify('NOTIFY_GET_PRODUCT_DETAILS', $product_id, $old_product);
```

Reply:
i am NOT a fan of backwards compatible in this situation.

MO, this is much ballyhoo over little.  and hardly a goat-rodeo.

this notifier was added to the repo by @torvista.  i think alerting him to its pending change should be enough.  i think he is competent and active enough to change any of his code accordingly.

my opinion is based on the level of activity of most plugins and their corresponding updates during the past 15 months.

the notifier in question `NOTIFY_GET_PRODUCT_DETAILS` is NOT in v158.  it got added for v158a.

so until said time, where any one can point me to released code that makes use of this notifier, and that said code is being actively used in the wild, my opinion is out with the old and in with the new.  

my opinion is subject to change with any new information.

finally, as someone who advocated for this change, i would like to thank chris for doing the heavy lifting on it. MO, this is a big improvement to the repo, and one that i completely endorse.

2nd finally, i would encourage all developers to now start using the product class as opposed to writing your own sql code for all your plugins.   again, i think this is a great improvement.

Reply:
How about we at least log a PHP warning if a customization is attempting to `attach` to that event?  That's in
https://github.com/zencart/zencart/blob/e3e94348e20e7075f8a728bfb4623129c145543f/includes/classes/traits/ObserverManager.php#L12-L31

I'd propose changing that section to
```php
trait ObserverManager
{
    static array $deprecatedNotifications = [
        'NOTIFY_GET_PRODUCT_DETAILS' => 'NOTIFY_GET_PRODUCT_OBJECT_DETAILS',
    ];

    /**
     * method used to an attach an observer to the notifier object
     *
     * NB. We have to get a little sneaky here to stop session based classes adding events ad infinitum
     * To do this we first concatenate the class name with the event id, as a class is only ever going to attach to an
     * event id once, this provides a unique key. To ensure there are no naming problems with the array key, we md5 the
     * unique name to provide a unique hashed key.
     *
     * @param object Reference to the observer class
     * @param array An array of eventId's to observe
     */
    function attach(&$observer, $eventIDArray)
    {
        $deprecated = array_keys(self::$deprecatedNotifications);
        foreach ($eventIDArray as $eventID) {
            if (in_array($eventID, $deprecated)) {
                trigger_error("Use of deprecated notification '$eventID'.  Consider using '" . self::$deprecatedNotifications[$eventID] . "' instead.", E_USER_WARNING);
                continue;
            }
            $nameHash = md5(get_class($observer) . $eventID);
            EventDto::getInstance()->setObserver($nameHash, array('obs' => &$observer, 'eventID' => $eventID));
        }
    }
```

Reply:
Having said that, the continuity of notifiers should be a basic promise of the product.   "Do it this way, debug it once, and it will work forever."   I hated telling people in 1.5.5 that they had to rework an observer, and hope to never have to do it again.

Reply:
#6473 should cover this ... or at least is the beginning of infrastructure to support marking deprecations.